### ISSUE TITLE
New major with JSX improvements

### ISSUE THREAD

**Issue by umbobabo** [Financial-Times/n-conversion-forms#276]
While migrating to JSX we can improve the codebase.
We should discuss what to use for self-documentation, demo, type checking, linting and other things we would like to have. Please share your ideas, then we can covert decisions into cards for this project https://github.com/Financial-Times/n-conversion-forms/projects/1

**Comment by umbobabo** [comment]
There are things I would like to have in this library:

1. A quick solution that doesn't require a lot of work for demo, playground, documentation. I never used https://github.com/styleguidist/react-styleguidist but it sounds less effort than storybook
2. Typescript
3. Prettier
4. Kill mocha, chai, sinon... and just keep Jest.

**Comment by ludrob** [comment]
1) demo: I know x-dash uses storybook, I haven't used styleguidist either but we could try a spike
2) why not, not sure about this, I often find it questionable/contrived to have typed languages on front-end. Especially in a well-structured team like ours with good practices, with tests and so on, the cost/benefit ratio is not that obvious
3) prettier: yes!
4) jest: yes!

**Comment by liamkeaton** [comment]
1. I've used Storybook in the past and it's served the purpose well. Not used anything else but hand rolled solutions so keen to hear about any others.
2. I would love to try using Typescript with JSDocs comments. This way you are writing pure JS and using JSDocs to define types. This way you can produce well-documented typed code along with type hinting, very useful for the utilities.
3. If it enforces consistency then I'm for it.
4. Never used Jest so it would be a learning curve for me. Could someone explain the benefits over the existing testing suite other than having just one package?

**Comment by ludrob** [comment]
> 4. Never used Jest so it would be a learning curve for me. Could someone explain the benefits over the existing testing suite other than having just one package?

You could say that jest is basically sinon/chai/mocha/istanbul combined. 
Everything is ready out of the box, it has all you need for mocking (no more proxyquire), testing and so on.
It's also incredibly quick compared to mocha (tests can run in parallel) and has better test isolation.
In terms of learning curve it's honestly mostly the same as mocha/chai/sinon, so you pretty much know 90% of it already.
In many cases you can literally copy over your mocha tests with only a handful of changes.
It generally feels well done, consistent and I find it quicker/easier to get stuff up and running.

**Comment by liamkeaton** [comment]
Any negatives we should be aware of?

**Comment by umbobabo** [comment]
In the past Jest has been criticised for the documentation that was quite bad. They fixed the issue in the last Major.

**Comment by remybach** [comment]
1. Yes! Storybook is fantastic (although I haven't gotten to use it very much, I worked on something that had very much the same principle).
2. Please... no Typescript 🙏. Agree with what Ludo said and on top of that, I don't believe this is the type of project that needs it. It would be adding complexity for no good reason. Much rather do as Liam suggested and just use JSDoc properly as it's far less like tying yourself to a steel ball and jumping in the ocean.
3. I'm willing to give prettier a good try.
4. I've started using Jest personally and it's been about on par with Mocha in terms of implementation (although, having fewer dependencies is nice). I could possibly be doing it inefficiently because I'm used to the Mocha way of doing it though.

**Comment by ludrob** [comment]
We've already covered most of the suggested changes on this issue.
Except that we still have some mocha tests.
I've raised https://financialtimes.atlassian.net/browse/ACQ-876 to cover this, so I'm closing this issue!
